doc/reference/kernel/other/fatal.rst

```diff
@@ -117,6 +117,26 @@ reports the failed test and its location, but lacks additional debugging
 information provided to assist the user in diagnosing the problem; its use is
 discouraged.
 
+__ASSERT_EXPR()
+---------------
+
+The ``__ASSERT_EXPR()`` macro is functionally equivalent to ``__ASSERT()`` if
```

Comment:
Would changing __ASSERT to behave like this instead of adding an additional macro lead to flash usage increase when asserts are disabled?

Putting expressions with side-effects in macros should be avoided, so I'm not sure I agree with adding an assert for this purpose. which is what the documentation below seems to indicate you could do.
For example `__ASSERT_EXPR(do_something() == 0, "something didn't happen");`

Reply:
IIUC the goal of this was not to support side effects, but to eliminate the need to conditionally declare variables/functions that are only referenced in `ASSERT` expressions; if the expression is discarded when asserts are disabled you get compiler warnings about unused variables, e.g:

    int condition = do_something();
    __ASSERT_EXPR(condition == 0, ...);  // this would remove the reference => warning while...
    (void)condition;                     // ...compiles cleanly
   

The non-enabled variant only does `(void)(expr);` so the message wouldn't be stored and the compiler should discard the expression.

Though I still don't think this is worth the hassle.

Reply:
The documentation seems to suggest that a use-case for this new macro is asserts with side-effects.

If the goal is to solve the unused variable then I think UNUSED_VARIABLE (similar to ARG_UNUSED) is a better idea.
Or alternatively define __ASSERT like __ASSERT_EXPR, if that does not increase the flash usage.